stdlib/public/core/StringObject.swift

```diff
@@ -193,34 +230,111 @@ extension _StringObject {
   static var _payloadMask: UInt {
     @inline(__always)
     get {
+      return 0x00FF_FFFF_FFFF_FFFF
+    }
+  }
+
+  @_versioned
+  @_inlineable
+  internal
+  var _variantBits: UInt {
+    @inline(__always)
+    get {
+      return UInt(truncatingIfNeeded: rawBits) & _StringObject._variantMask
+    }
+  }
+#endif // arch(i386) || arch(arm)
+
+  @_versioned
+  @_inlineable
+  internal
+  var referenceBits: UInt {
+    @inline(__always)
+    get {
 #if arch(i386) || arch(arm)
-      return 0x0FFF_FFFF
+      guard case let .strong(object) = _variant else { return 0 }
+      return Builtin.reinterpretCast(object)
 #else
-      return 0x00FF_FFFF_FFFF_FFFF
+      return _bitPattern(_object) & UInt(_StringObject._payloadMask)
+#endif
+    }
+  }
+
+  @_versioned
+  @_inlineable
+  internal
+  var payloadBits: UInt {
+    @inline(__always)
+    get {
+#if arch(i386) || arch(arm)
+      if case .strong(_) = _variant { return 0 }
```

Comment:
Shouldn't the return value be the reference? I suppose all uses only do this for value-Strings, but if we want to not return the address we should comment and sanityCheck on the variant.

Reply:
I think sanityCheck is the better option; we have `referenceBits` for accessing the pointer value.